table/tables/tables.go

```diff
@@ -688,20 +668,64 @@ func (t *TableCommon) AddRecord(ctx sessionctx.Context, r []types.Datum, opts ..
 		return nil, err
 	}
 	value := writeBufs.RowValBuf
-	if err = execBuf.Set(key, value); err != nil {
+
+	var keyFlags kv.KeyFlags
+	var ctx context.Context
+	if opt.Ctx != nil {
+		ctx = opt.Ctx
+	} else {
+		ctx = context.Background()
+	}
+	skipCheck := sctx.GetSessionVars().StmtCtx.BatchCheck
+	if (t.meta.IsCommonHandle || t.meta.PKIsHandle) && !skipCheck && !opt.SkipHandleCheck {
+		if sctx.GetSessionVars().PresumeKeyNotExists {
+			var v []byte
+			v, err = txn.GetMemBuffer().Get(ctx, key)
+			if err != nil {
+				keyFlags = keyFlags.MarkPresumeKeyNotExists()
+			}
+			if err == nil && len(v) == 0 {
+				err = kv.ErrNotExist
```

Comment:
maybe this branch also need `keyFlags = keyFlags.MarkPresumeKeyNotExists()`~?

Reply:
This branch is for tombstones, `TxnState.Get` will handle them by these steps:
1. If the tombstone is in `stmtBuf`, return `kv.ErrNotExist` without visiting `tikvTxn.us.MemBuffer`;
	https://github.com/pingcap/tidb/blob/75b254c8ebf21483f591900638e7a37851764d85/session/txn.go#L294-L306
2. If the tombstone is in `tikvTxn.us.MemBuffer`, return `kv.ErrNotExist` without access `us.snapshot`;
	https://github.com/pingcap/tidb/blob/75b254c8ebf21483f591900638e7a37851764d85/kv/union_store.go#L104-L120

Let's see what will happen if we treat tombstone as **key does not exist** 🤣

```sql
delete from t where a = 1;
Query OK, 1 row affected (0.001 sec)

insert into t value (1);
Query OK, 1 row affected (0.001 sec)

commit;
ERROR 1062 (23000): Duplicate entry '1' for key 'PRIMARY'
```

Reply:
This one reason for this refactor, I think it is better to handle `delete tombstone` explicit rather than convert it to `kv.ErrNotExist`.

Reply:
ok, I get it~ :laughing: